src/libraries/Microsoft.Bcl.TimeProvider/src/System/Threading/Tasks/TimeProviderTaskExtensions.cs

```diff
@@ -0,0 +1,171 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Threading.Tasks
+{
+    public static class TimeProviderTaskExtensions
+    {
+        private sealed class DelayState : TaskCompletionSource<bool>
+        {
+            public DelayState() : base(TaskCreationOptions.RunContinuationsAsynchronously) {}
+            public ITimer Timer { get; set; }
+            public CancellationTokenRegistration Registration { get; set; }
+        }
+
+        public static Task Delay(this TimeProvider timeProvider, TimeSpan delay, CancellationToken cancellationToken = default)
+        {
+            if (timeProvider == TimeProvider.System)
+            {
+                return Task.Delay(delay, cancellationToken);
+            }
+
+            if (timeProvider is null)
+            {
+                throw new ArgumentNullException(nameof(timeProvider));
+            }
+
+            if (delay != Timeout.InfiniteTimeSpan && delay < TimeSpan.Zero)
+            {
+                throw new ArgumentOutOfRangeException(nameof(delay));
+            }
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return Task.FromCanceled(cancellationToken);
+            }
+
+            if (delay == TimeSpan.Zero)
+            {
+                return Task.CompletedTask;
+            }
+
+            DelayState state = new();
+            state.Timer = timeProvider.CreateTimer(delayState =>
+            {
+                DelayState s = (DelayState)delayState;
+                s.TrySetResult(true);
+                s.Registration.Dispose();
+                s.Timer.Dispose();
+            }, state, Timeout.InfiniteTimeSpan, Timeout.InfiniteTimeSpan);
+
+            state.Timer.Change(delay, Timeout.InfiniteTimeSpan);
```

Comment:
Probably worth a comment highlighting why we're first creating it as Infinite and then subsequently calling Change (to avoid the race condition on storing the timer into state.Timer and the timer firing and seeing s.Timer as null).